### DISCUSSION TITLE
Zfinx support in riscv-arch-test added

### DISCUSSION THREAD

**Discussion by mikaelsky** [stnolting/neorv32#838] [Show and tell]
@stnolting its taken me a while to get time to add support for this. This is an upgraded version of what I've used internally to validate that the Zfinx extension is compliant. Upgraded as in it converts everything and has room for subnormal flush support.

I've added initial support for Zfinx to the riscv-arch-test suite. There is a branch here that has Zfinx test suite support:
[https://github.com/mikaelsky/riscv-arch-test/tree/Zfinx_test_suite](https://github.com/mikaelsky/riscv-arch-test/tree/Zfinx_test_suite)

The changes are: 
- Adding specific Zfinx test functions into test_macros.h. I've only tested the neorv32 subset of the full F test suite, but should generate the remainder of the Zfinx tests correctly.
- Adding 2 scripts for converting F test to Zfinx tests and generating test suites.
Note: I've yet to add support for flushing sub-normals in the ref simulation, so quite a few test will fail as the ref will correctly interpret sub-normals.

The branch doesn't include any tests as the scripts are tunable as to what is generated. This to allow us to generate neorv32 compliant zfinx test suites. Basically you will need to manually run the scripts first before calling the riscof framework.

There is also a branch of neorv32-riscof: [https://github.com/mikaelsky/neorv32-riscof/tree/Zfinx_test_suite](https://github.com/mikaelsky/neorv32-riscof/tree/Zfinx_test_suite)

The changes are:
- Expanding the ISA definition in neorv32_isa.yaml to include Zfinx.
- Added a Zfinx ISA decode path to riscof_neorv32.py and riscof_sail_cSim.py
- Enabled zfinx for the core instantiated in the testbench

To generate a test suite simply go into the Zfinx test suite directory. Create a src folder if it doesn't exists and run: `./script/generate_test_suite.py ../F/ ./` this will generate a neorv32 compliant test suite.
Note: All the neorv32 test run and generate an output with GHDL.
Note: I recommend keeping an eye on Sail as Zfinx support is still a work-in-progress. If test logs begin to go off the rails said instruction isn't working right. What I do is keep htop running in a terminal and if a given test takes a long time something is off.
Note: calling either script with -h or no parameters will output all the options available.

To just run the Zfinx test the following can be run from the root of the riscv32-riscvof folder: `riscof --verbose info run --config ./config.ini --suite ./riscv-arch-test/riscv-test-suite/rv32i_m/Zfinx --env ./riscv-arch-test/riscv-test-suite/env`. This is quite useful to just check a singular extension.

**Comment by mikaelsky** [comment]
After a bit of script tuning to ensure we don't clobber the result address register the neorv32 zfinx suite now runs. There are still ... challenges .. with Sail Zfinx support. I'm unsure what version of Sail we are running. I can see it "takes" the instruction and does - something - but if you look at the signature file something is off e,g the fclass ref signature is borked. All the test fail compare, when in reality at least half should pass even with subnormals in the test suite.

For my internal testing I ran the F test suite on sail and just copied over the signature files. The ref signature files where compared with the neorv32 running the converted zfinx test suite.

**Comment by mikaelsky** [comment]
Result of my local run on my home server:

![image](https://github.com/stnolting/neorv32/assets/119539842/ea906339-a02b-4a9b-bf66-5737e4a22989)

**Reply by stnolting** [reply]
Did you find out which version of SAIL you are using? Did you build it by your own or did you use some prebuild binaries?

**Reply by mikaelsky** [reply]
I just used the prebuilt that lives in neorv32-riscof.

**Reply by mikaelsky** [reply]
When I look at the log out from Sail its seems funky. Basically it seems like everything is rotated left by exactly 1 byte. More debugging is probably needed. Looking at the log output I'm tempted to make a test harness for the RVVI that prints out a log with the delta changes for every instruction run, which seems to be what is in the Sail log. Would allow some crude comparison of sail vs neorv32.

**Comment by stnolting** [comment]
> There is also a branch of neorv32-riscof: https://github.com/mikaelsky/neorv32-riscof/tree/Zfinx_test_suite

That's really great! I tried to come up with something like this myself, but I failed miserably 😅. I haven't checked all the details yet, but I wonder how close this is to the current Zfinx integration proposal: https://github.com/riscv-non-isa/riscv-arch-test/pull/367

**Reply by mikaelsky** [reply]
From what I can see the PR is a bit more encompassent, even though I suspect all they are doing is a more "elegant" version of my quick hack :) Ideally they would generate a new .cgt file that then generates the tests.
What I have is a bit simpler and that all I do is transform the rv32m_I/F test into Zfinx tests as one can argue that Zfinx is just a subset of F.
The scripts I made for this one have more coverage generation than the internal one I did last summer.

When/if the Zfinx intergration gets approved I would pivot that, but I couldn't wait as I needed a tool to determine the quality of the FPU :)

**Comment by stnolting** [comment]
FYI

`Zfinx` support has just been added to riscv-arch-test: https://github.com/riscv-non-isa/riscv-arch-test/releases/tag/3.8.11 🚀

I don't think it will work right away, but let's see...

**Reply by mikaelsky** [reply]
Interesting, let me take a look at this instead of using our internal "hack".
